pkg/download/buffer.go

```diff
@@ -93,51 +122,52 @@ func (m *BufferMode) fileToBuffer(ctx context.Context, url string) (*bytes.Buffe
 	if numChunks <= 0 {
 		numChunks = 1
 	}
-	if numChunks > m.maxChunks() {
-		numChunks = m.maxChunks()
+	if numChunks > m.maxConcurrency() {
+		numChunks = m.maxConcurrency()
 	}
 
+	readersCh := make(chan io.Reader, m.maxConcurrency()+1)
+	readersCh <- br
+
+	startOffset := m.minChunkSize()
+
 	chunkSize := remainingBytes / int64(numChunks)
 	if chunkSize < 0 {
-		firstChunkResp.Body.Close()
 		return nil, -1, fmt.Errorf("error: chunksize incorrect - result is negative, %d", chunkSize)
 	}
 
-	logger.Debug().Str("url", url).
-		Int64("size", fileSize).
-		Int("connections", numChunks).
-		Int64("chunkSize", chunkSize).
-		Msg("Downloading")
-
-	errGroup, ctx := errgroup.WithContext(ctx)
-	errGroup.Go(func() error {
-		return m.downloadChunk(firstChunkResp, data[0:m.minChunkSize()])
-	})
-
-	startOffset := m.minChunkSize()
+	m.queue.submit(func() {
```

Comment:
I did try doing this but there's quite a lot of data that the closure closes over that would have to be passed in some other way and it made things more complicated rather than simpler.